File: sdk/nodejs/src/repository.ts
import {Entity} from "./entity";
import {Container} from "./container";
import {BooleanExpression, Event} from "./model/extension";
import {Watcher} from "./watcher";
import {ListRecordParams} from "./list-record-params";

export interface Repository<T extends Entity> {
    create(record: T): Promise<T>;

    get(id: string, resolveReferences?: string[]): Promise<T>;

    update(record: Partial<T> & Entity): Promise<T>;

    delete(id: string): Promise<T>;

    apply(record: Partial<T>): Promise<T>;

    load(record: Partial<T>, resolveReferences?: string[]): Promise<T>;

    list(params?: ListRecordParams): Promise<Container<T>>;

    watch(eventConsumer: (event: Event) => void): Watcher<T>;
}

File: sdk/java/src/main/java/io/apibrew/client/Repository.java
package io.apibrew.client;

import io.apibrew.client.model.Extension;

import java.util.List;
import java.util.function.Consumer;

public interface Repository<T extends Entity> {

    T create(T record);
    T get(String id);
    T get(String id, List<String> resolveReferences);
    T update(T record);
    T delete(String id);
    T apply(T record);
     Container<T> list();
     Container<T> list(ListRecordParams params);

    Container<T> list(Extension.BooleanExpression ...query);

     T load(T record);

    Watcher<T> watch(Consumer<Extension.Event> eventConsumer);
}
